python_modules/dagster/dagster/core/asset_defs/assets.py

```diff
@@ -428,6 +451,69 @@ def with_resources(self, resource_defs: Mapping[str, ResourceDefinition]) -> "As
             group_names=self._group_names,
         )
 
+    def without_resources(self) -> "AssetsDefinition":
+        # In the case where we provided a system-created resource key, we need
+        # to remove it without requiring users provide it later.
+        if len(self.asset_keys) == 1 and isinstance(self.node_def, OpDefinition):
+
+            io_manager_key = io_manager_key_for_asset_key(self.asset_key)
+
+            op_def = self.op
+            output_def = op_def.output_defs[0]
+            if output_def.io_manager_key == io_manager_key_for_asset_key(self.asset_key):
+                io_manager_key = "io_manager"
+                new_required_keys = {
+                    io_manager_key,
+                    *{
+                        key
+                        for key in op_def.required_resource_keys
+                        if key != io_manager_key_for_asset_key(self.asset_key)
+                    },
+                }
+
+            else:
+                io_manager_key = output_def.io_manager_key
+                new_required_keys = op_def.required_resource_keys
+
+            output_def = OutputDefinition(
+                dagster_type=output_def.dagster_type,
+                name=output_def.name,
+                description=output_def.description,
+                is_required=output_def.is_required,
+                io_manager_key=io_manager_key,
+                metadata=output_def.metadata,
+                asset_key=output_def._asset_key,
+                asset_partitions=output_def._asset_partitions_fn,
+                asset_partitions_def=output_def._asset_partitions_def,
+            )
+
+            node_def = OpDefinition(
+                name=op_def.name,
```

Comment:
In practice, I think this will be ok, but I'm a little wary of this sort of implementation (because when you do this, you end up with two conflicting OpDefinitions with the same name).

This only mattes once you put this asset on a repository, which you should never be doing if I'm understanding this correctly, but still.

Also seems like a pain / fairly brittle to do all this copying.

Can we do something fancier here, where we keep around a list of keys for the io manager defs that were removed? something like self._removed_io_manager_keys

Then when we call with_resources, we can map whatever resource_def is provided for key "io_manager" to all of the "self._removed_io_manager_keys".

kinda ugly but maybe there's a way to riff on that which feels more elegant. just seems like in the future if a field is added to any of OpDefinition, AssetsDefinition, or OutputDefinition, this code could potentially break

Reply:
i guess this would only work if we expect the user to invoke with_resources to add resources back on (which I'm not sure is the case)

Reply:
While I think it's probably fine that the op has the same name, and we could add an optional name parameter as an escape hatch, totally agreed that this implementation is pretty jank. I like your idea of keeping around some sort of reference to the old key, it feels a bit less invasive.